lib/usdUfe/ufe/UsdUndoDuplicateSelectionCommand.cpp

```diff
@@ -161,47 +161,72 @@ Ufe::SceneItemList UsdUndoDuplicateSelectionCommand::targetItems() const
+    // 3. A property references a prim that's in neither set.
+    //        -> Delete the reference.
+    for (size_t i = 0; i < referencedPaths.size(); ++i) {
+        const auto& referencedPath = referencedPaths[i];
+
+        // If the referenced path points to a prim in the duplicate set, there's nothing to do.
+        const bool isInDuplicateSet = referencedPath.HasPrefix(duplicatePrimPath);
+        if (isInDuplicateSet) {
+            continue;
         }
-        const auto endPath = otherPairs.upper_bound(finalPath);
-        bool       isExternalPath = true;
-        for (; itPath != endPath; ++itPath) {
-            if (*itPath == duplicatePair) {
-                // That one was correctly processed by USD when duplicating.
-                isExternalPath
-                    = !finalPath.HasPrefix(itPath->first) && !finalPath.HasPrefix(itPath->second);
-                continue;
-            }
-            finalPath = finalPath.ReplacePrefix(itPath->first, itPath->second);
-            if (path != finalPath) {
-                pathVec[i] = finalPath;
-                hasChanged = true;
-                isExternalPath = false;
-                break;
-            }
+
+        // Check if the original set contains the referenced path. This is true if any path in the
+        // DuplicatePathsMap is a prefix of the referenced path.
+        //
+        // Since paths are ordered lexicographically, a prefix of a path is always less then or
+        // equal to the path itself. Also, the prefix will always be "close to" the path itself.
+        // Thus, we can get away with checking only a single candidate: The last path thats less
+        // than or equal to the reference path.
+        //
+        // `upper_bound()` returns the first path that's greater than the referenced path. Our
+        // candidate is the path before that.
+        const auto firstLargerPath = allPairs.upper_bound(referencedPath);
+        if (firstLargerPath == allPairs.begin()) {
+            // All paths in the source set are greater, so none of them can be a prefix.
+            indicesToRemove.push_front(i);
+            continue;
         }
-        if (isExternalPath) {
-            hasChanged = true;
+
+        // Check if our candidate is a prefix of the referenced path.
+        const auto lastSmallerPath = std::prev(firstLargerPath);
+        const bool isInOriginalSet = referencedPath.HasPrefix(lastSmallerPath->first);
+        if (!isInOriginalSet) {
             indicesToRemove.push_front(i);
+            continue;
         }
+
+        // Update the path to point to the respective path in the duplicate set.
+        referencedPath.ReplacePrefix(lastSmallerPath->first, lastSmallerPath->second);
+        referencedPathsChanged = true;
     }
+
     for (size_t toRemove : indicesToRemove) {
-        pathVec.erase(pathVec.cbegin() + toRemove);
+        referencedPaths.erase(referencedPaths.cbegin() + toRemove);
```

Comment:
For clarity, would be good to mention that this works because the indices are in reverse order due to the use of `push_front`. This means no need to adjust because the vector has closed the gap of a deleted item.